src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorParsingPhase.cs

```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Razor.Evolution
+{
+    internal class DefaultRazorParsingPhase : RazorEnginePhaseBase, IRazorParsingPhase
+    {
+        protected override void ExecuteCore(RazorCodeDocument codeDocument)
+        {
+            var syntaxTree = RazorSyntaxTree.Parse(codeDocument.Source);
```

Comment:
So if a user wants to provide their own SyntaxTree to work on they need to remove the parsing phase from the builder. Might be worth trying to retrieve the syntax tree and then parsing if there wasn't one already provided.